src/engine/platform/mixer.cpp

```diff
@@ -49,63 +47,32 @@ struct AudioStruct {
-    ASSERT(delay_time < CHANNEL_BUFFER_LENGTH_SECONDS, "Delay time is longer than channel buffer");
-    auto delay_func = [] (Effect *effect, f32 *buffer, u32 start, u32 len) -> void {
-        if (effect->delay.delay_len_seconds != effect->delay._prev_delay_len_seconds) {
+void Channel::effect(u32 start, u32 len) {
+    if (delay.active) {
+        if (delay.len_seconds != delay._prev_len_seconds) {
             //TODO(GS) crackling when length is changed while sound is playing
-            effect->delay.delay_len = (u32) AUDIO_SAMPLE_RATE * effect->delay.delay_len_seconds * 2;
-            effect->delay._prev_delay_len_seconds = effect->delay.delay_len_seconds;
+            delay.len = (u32) AUDIO_SAMPLE_RATE * delay.len_seconds * 2;
+            delay._prev_len_seconds = delay.len_seconds;
         }
         for (u32 i = 0; i < len; i++) {
             u32 cur_pos = (start + i) % CHANNEL_BUFFER_LENGTH;
-            u32 prev_pos = (start + i - effect->delay.delay_len + CHANNEL_BUFFER_LENGTH) % CHANNEL_BUFFER_LENGTH;
-            buffer[cur_pos] += buffer[prev_pos] * effect->delay.feedback;
+            u32 pre_pos = (start + i - delay.len + CHANNEL_BUFFER_LENGTH) % CHANNEL_BUFFER_LENGTH;
+            buffer[cur_pos] += buffer[pre_pos] * delay.feedback;
         }
-    };
-    Effect effect = {};
-    effect.id = invalid_id();
-    effect.effect = delay_func;
-    effect.delay.feedback = feedback;
-    effect.delay.delay_len_seconds = delay_time;
-    return effect;
-}
-
-EffectID add_effect(Effect effect, u32 channel) {
-    ASSERT(channel < NUM_CHANNELS, "Invalid channel");
-    for (u32 i = 0; i < NUM_EFFECTS; i++) {
-        if (audio_struct.effects[channel][i].effect) continue;
-        EffectID id = {(s32) channel, i, audio_struct.num_effects};
-        effect.id = id;
-        audio_struct.effects[effect.id.channel][effect.id.slot] = effect;
-        audio_struct.num_effects++;
-        return effect.id;
     }
-    ERR("Not enough free effect slots on channel %d, skipping effect", channel);
-    return invalid_id();
 }
 
-Effect *fetch_effect(EffectID id) {
-    if (!audio_struct.effects[id.channel][id.slot].effect) {
-        return nullptr;
-    }
-    return &audio_struct.effects[id.channel][id.slot];
-}
-
-bool remove_effect(EffectID id) {
-    if (!audio_struct.effects[id.channel][id.slot].effect) {
-        ERR("Invalid EffectID");
-        return false;
-    }
-    audio_struct.effects[id.channel][id.slot].effect = nullptr;
-    return true;
+void Channel::set_delay(f32 feedback, f32 len_seconds) {
+    delay.active = true;
+    delay.feedback = feedback;
+    delay.len_seconds = len_seconds;
 }
 
-void clear_effects(u32 channel) {
-    ASSERT(channel < NUM_CHANNELS, "Invalid channel");
-    for (u32 i = 0; i < NUM_EFFECTS; i++) {
-        audio_struct.effects[channel][i].effect = nullptr;
+Channel *fetch_channel(u32 channel_id) {
+    if (channel_id >= NUM_CHANNELS) {
+        return nullptr;
     }
```

Comment:
You can probably assert this, since the number of channels is constant and known. Returning nullptr requires people to check for it, which quickly gets annoying.